File: frontend/src/services/autoPilot.ts
import { api } from "@/lib/api";

export interface AutoPilotPlan {
  id?: string;
  name: string;
  description: string;
  frequency: 'daily' | 'weekly' | 'biweekly' | 'monthly' | 'quarterly';
  startDate: Date;
  endDate?: Date;
  isActive: boolean;
  questionRotation: boolean;
  reminderSettings: ReminderSettings;
  distributionChannels: string[];
  targetAudience: string[];
  maxResponses?: number;
  createdAt?: string;
  updatedAt?: string;
}

export interface ReminderSettings {
  enabled: boolean;
  frequency: 'daily' | 'weekly' | 'biweekly';
  maxReminders: number;
  delayDays: number;
  messageTemplate: string;
  excludeResponded: boolean;
  autoCloseAfterDays: number;
  reminderDays: number[];
}

export interface AutoPilotSurvey {
  id?: string;
  planId: string;
  surveyId: string;
  scheduledDate: Date;
  sentDate?: Date;
  status: 'scheduled' | 'sent' | 'completed' | 'failed';
  questionSet: QuestionSet;
  reminderCount: number;
  lastReminderDate?: Date;
  responseCount: number;
  targetCount: number;
}

export interface QuestionSet {
  id: string;
  name: string;
  questions: Question[];
  category: string;
  rotationIndex: number;
}

export interface Question {
  id: string;
  text: string;
  category: string;
  order: number;
  required: boolean;
  variations?: string[];
  currentVariation?: number;
}

export interface AutoPilotMetrics {
  totalSurveys: number;
  activePlans: number;
  averageResponseRate: number;
  totalResponses: number;
  nextScheduledSurvey?: Date;
  recentActivity: AutoPilotActivity[];
}

export interface AutoPilotActivity {
  id: string;
  type: 'survey_sent' | 'reminder_sent' | 'plan_created' | 'plan_updated';
  description: string;
  timestamp: Date;
  planId?: string;
  surveyId?: string;
}

class AutoPilotService {
  // Auto-Pilot Plan Management
  async createPlan(plan: AutoPilotPlan): Promise<AutoPilotPlan> {
    try {
      const response = await api.post('/auto-pilot/plans', plan);
      return response.data;
    } catch (error) {
      console.error('Error creating auto-pilot plan:', error);
      throw error;
    }
  }

  async getPlans(): Promise<AutoPilotPlan[]> {
    try {
      const response = await api.get('/auto-pilot/plans');
      return response.data;
    } catch (error) {
      console.error('Error getting auto-pilot plans:', error);
      throw error;
    }
  }

  async getPlan(id: string): Promise<AutoPilotPlan> {
    try {
      const response = await api.get(`/auto-pilot/plans/${id}`);
      return response.data;
    } catch (error) {
      console.error('Error getting auto-pilot plan:', error);
      throw error;
    }
  }

  async updatePlan(id: string, plan: Partial<AutoPilotPlan>): Promise<AutoPilotPlan> {
    try {
      const response = await api.put(`/auto-pilot/plans/${id}`, plan);
      return response.data;
    } catch (error) {
      console.error('Error updating auto-pilot plan:', error);
      throw error;
    }
  }

  async deletePlan(id: string): Promise<void> {
    try {
      await api.delete(`/auto-pilot/plans/${id}`);
    } catch (error) {
      console.error('Error deleting auto-pilot plan:', error);
      throw error;
    }
  }

  async activatePlan(id: string): Promise<AutoPilotPlan> {
    try {
      const response = await api.post(`/auto-pilot/plans/${id}/activate`);
      return response.data;
    } catch (error) {
      console.error('Error activating auto-pilot plan:', error);
      throw error;
    }
  }

  async deactivatePlan(id: string): Promise<AutoPilotPlan> {
    try {
      const response = await api.post(`/auto-pilot/plans/${id}/deactivate`);
      return response.data;
    } catch (error) {
      console.error('Error deactivating auto-pilot plan:', error);
      throw error;
    }
  }

  // Auto-Pilot Survey Management
  async getScheduledSurveys(planId?: string): Promise<AutoPilotSurvey[]> {
    try {
      const params = planId ? { planId } : {};
      const response = await api.get('/auto-pilot/surveys', { params });
      return response.data;
    } catch (error) {
      console.error('Error getting scheduled surveys:', error);
      throw error;
    }
  }

  async getSurvey(id: string): Promise<AutoPilotSurvey> {
    try {
      const response = await api.get(`/auto-pilot/surveys/${id}`);
      return response.data;
    } catch (error) {
      console.error('Error getting auto-pilot survey:', error);
      throw error;
    }
  }

  async updateSurvey(id: string, survey: Partial<AutoPilotSurvey>): Promise<AutoPilotSurvey> {
    try {
      const response = await api.put(`/auto-pilot/surveys/${id}`, survey);
      return response.data;
    } catch (error) {
      console.error('Error updating auto-pilot survey:', error);
      throw error;
    }
  }

  async cancelSurvey(id: string): Promise<void> {
    try {
      await api.post(`/auto-pilot/surveys/${id}/cancel`);
    } catch (error) {
      console.error('Error canceling auto-pilot survey:', error);
      throw error;
    }
  }

  // Question Rotation Management
  async getQuestionSets(category?: string): Promise<QuestionSet[]> {
    try {
      const params = category ? { category } : {};
      const response = await api.get('/auto-pilot/question-sets', { params });
      return response.data;
    } catch (error) {
      console.error('Error getting question sets:', error);
      throw error;
    }
  }

  async createQuestionSet(questionSet: QuestionSet): Promise<QuestionSet> {
    try {
      const response = await api.post('/auto-pilot/question-sets', questionSet);
      return response.data;
    } catch (error) {
      console.error('Error creating question set:', error);
      throw error;
    }
  }

  async rotateQuestions(planId: string): Promise<QuestionSet> {
    try {
      const response = await api.post(`/auto-pilot/plans/${planId}/rotate-questions`);
      return response.data;
    } catch (error) {
      console.error('Error rotating questions:', error);
      throw error;
    }
  }

  // Reminder System
  async sendReminder(surveyId: string): Promise<void> {
    try {
      await api.post(`/auto-pilot/surveys/${surveyId}/send-reminder`);
    } catch (error) {
      console.error('Error sending reminder:', error);
      throw error;
    }
  }

  async getReminderHistory(surveyId: string): Promise<any[]> {
    try {
      const response = await api.get(`/auto-pilot/surveys/${surveyId}/reminders`);
      return response.data;
    } catch (error) {
      console.error('Error getting reminder history:', error);
      throw error;
    }
  }

  async updateReminderSettings(planId: string, settings: ReminderSettings): Promise<AutoPilotPlan> {
    try {
      const response = await api.put(`/auto-pilot/plans/${planId}/reminder-settings`, settings);
      return response.data;
    } catch (error) {
      console.error('Error updating reminder settings:', error);
      throw error;
    }
  }

  // Metrics and Analytics
  async getMetrics(): Promise<AutoPilotMetrics> {
    try {
      const response = await api.get('/auto-pilot/metrics');
      return response.data;
    } catch (error) {
      console.error('Error getting auto-pilot metrics:', error);
      throw error;
    }
  }

  async getActivityLog(planId?: string, limit?: number): Promise<AutoPilotActivity[]> {
    try {
      const params: any = {};
      if (planId) params.planId = planId;
      if (limit) params.limit = limit;
      
      const response = await api.get('/auto-pilot/activity', { params });
      return response.data;
    } catch (error) {
      console.error('Error getting activity log:', error);
      throw error;
    }
  }

  // Scheduling Logic
  calculateNextSurveyDate(plan: AutoPilotPlan, lastSurveyDate?: Date): Date {
    const baseDate = lastSurveyDate || plan.startDate;
    const frequency = plan.frequency;
    
    switch (frequency) {
      case 'daily':
        return new Date(baseDate.getTime() + 24 * 60 * 60 * 1000);
      case 'weekly':
        return new Date(baseDate.getTime() + 7 * 24 * 60 * 60 * 1000);
      case 'biweekly':
        return new Date(baseDate.getTime() + 14 * 24 * 60 * 60 * 1000);
      case 'monthly':
        return new Date(baseDate.getTime() + 30 * 24 * 60 * 60 * 1000);
      case 'quarterly':
        return new Date(baseDate.getTime() + 90 * 24 * 60 * 60 * 1000);
      default:
        return new Date(baseDate.getTime() + 7 * 24 * 60 * 60 * 1000);
    }
  }

  calculateReminderDate(surveyDate: Date, reminderDay: number): Date {
    return new Date(surveyDate.getTime() + reminderDay * 24 * 60 * 60 * 1000);
  }

  shouldSendReminder(survey: AutoPilotSurvey, settings: ReminderSettings): boolean {
    if (!settings.enabled) return false;
    if (survey.reminderCount >= settings.maxReminders) return false;
    if (survey.responseCount >= survey.targetCount * 0.8) return false; // 80% response rate threshold
    
    const surveyDate = survey.sentDate;
    if (!surveyDate) return false;
    
    // Check if it's time for the next reminder based on predefined days
    const reminderDays = settings.reminderDays || [3, 7]; // Default to Day 3 and Day 7 reminders
    const currentDay = Math.floor((new Date().getTime() - new Date(surveyDate).getTime()) / (24 * 60 * 60 * 1000));
    
    // Check if we should send a reminder today
    const shouldSendToday = reminderDays.includes(currentDay) && 
                           survey.reminderCount < reminderDays.indexOf(currentDay) + 1;
    
    return shouldSendToday;
  }

  shouldCloseSurvey(survey: AutoPilotSurvey, settings: ReminderSettings): boolean {
    if (!settings.enabled) return false;
    
    const surveyDate = survey.sentDate;
    if (!surveyDate) return false;
    
    // Close survey based on auto-close setting
    const autoCloseDays = settings.autoCloseAfterDays || 10; // Default to 10 days
    const daysSinceSent = Math.floor((new Date().getTime() - new Date(surveyDate).getTime()) / (24 * 60 * 60 * 1000));
    return daysSinceSent >= autoCloseDays;
  }

  // Question Rotation Logic
  generateQuestionVariations(question: Question): string[] {
    if (question.variations && question.variations.length > 0) {
      return question.variations;
    }
    
    // Generate variations based on question category
    const variations: { [key: string]: string[] } = {
      satisfaction: [
        "How satisfied are you with your current role?",
        "Overall, how satisfied are you with your job right now?",
        "How positive is your daily experience at work?",
        "Do you feel good about coming to work each day?",
        "How content are you with your current responsibilities?"
      ],
      enps: [
        "How likely are you to recommend working here to a friend?",
        "Would you tell a friend this is a good place to work?",
        "How likely are you to encourage others to join this company?",
        "If a friend was job-hunting, would you suggest they apply here?",
        "Would you speak positively about working here to others?"
      ],
      manager: [
        "Do you feel supported by your manager?",
        "How well does your manager help you succeed?",
        "Do you trust your manager's leadership?",
        "Does your manager listen to your ideas and concerns?",
        "How approachable is your manager when you need help?"
      ],
      collaboration: [
        "How connected do you feel with your teammates?",
        "Do you feel like you belong in your team?",
        "How well do you and your teammates work together?",
        "Do your colleagues support you when needed?",
        "How good is the teamwork in your department?"
      ]
    };
    
    return variations[question.category] || [question.text];
  }

  rotateQuestionText(question: Question): string {
    const variations = this.generateQuestionVariations(question);
    const currentIndex = question.currentVariation || 0;
    const nextIndex = (currentIndex + 1) % variations.length;
    
    return variations[nextIndex];
  }

  // Validation
  validatePlan(plan: AutoPilotPlan): { isValid: boolean; errors: string[] } {
    const errors: string[] = [];

    if (!plan.name?.trim()) {
      errors.push('Plan name is required');
    }

    if (!plan.frequency) {
      errors.push('Frequency is required');
    }

    if (!plan.startDate) {
      errors.push('Start date is required');
    }

    if (plan.endDate && plan.startDate >= plan.endDate) {
      errors.push('End date must be after start date');
    }

    if (!plan.distributionChannels?.length) {
      errors.push('At least one distribution channel is required');
    }

    if (!plan.targetAudience?.length) {
      errors.push('Target audience is required');
    }

    if (plan.reminderSettings.enabled) {
      if (plan.reminderSettings.maxReminders < 1) {
        errors.push('Maximum reminders must be at least 1');
      }
      if (plan.reminderSettings.autoCloseAfterDays < 1) {
        errors.push('Auto-close days must be at least 1');
      }
      if (!plan.reminderSettings.reminderDays?.length) {
        errors.push('Reminder days must be specified');
      }
    }

    return {
      isValid: errors.length === 0,
      errors
    };
  }

  // Local Storage for offline functionality
  savePlanToLocalStorage(plan: AutoPilotPlan): void {
    try {
      const plans = this.getPlansFromLocalStorage();
      const existingIndex = plans.findIndex(p => p.id === plan.id);
      
      if (existingIndex >= 0) {
        plans[existingIndex] = { ...plan, updatedAt: new Date().toISOString() };
      } else {
        plans.push({ ...plan, id: `local_${Date.now()}`, createdAt: new Date().toISOString() });
      }
      
      localStorage.setItem('novora_auto_pilot_plans', JSON.stringify(plans));
    } catch (error) {
      console.error('Error saving plan to local storage:', error);
    }
  }

  getPlansFromLocalStorage(): AutoPilotPlan[] {
    try {
      const data = localStorage.getItem('novora_auto_pilot_plans');
      return data ? JSON.parse(data) : [];
    } catch (error) {
      console.error('Error reading plans from local storage:', error);
      return [];
    }
  }

  // Sync local plans with server
  async syncLocalPlans(): Promise<void> {
    try {
      const localPlans = this.getPlansFromLocalStorage();
      const unsyncedPlans = localPlans.filter(p => p.id?.startsWith('local_'));
      
      for (const plan of unsyncedPlans) {
        try {
          await this.createPlan(plan);
          // Remove from local storage after successful sync
          const plans = this.getPlansFromLocalStorage();
          const filteredPlans = plans.filter(p => p.id !== plan.id);
          localStorage.setItem('novora_auto_pilot_plans', JSON.stringify(filteredPlans));
        } catch (error) {
          console.error(`Error syncing plan ${plan.id}:`, error);
        }
      }
    } catch (error) {
      console.error('Error syncing local plans:', error);
    }
  }
}

export const autoPilotService = new AutoPilotService();

File: backend/app/api/v1/auto_pilot.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime, timedelta
from app.core.database import get_db
from app.api.deps import get_current_user
from app.models.base import User
from app.models.advanced import AutoPilotPlan, AutoPilotSurvey, QuestionBank
from app.schemas.auto_pilot import (
    AutoPilotPlanCreate,
    AutoPilotPlanUpdate,
    AutoPilotPlanResponse,
    AutoPilotSurveyResponse,
    AutoPilotMetricsResponse,
    AutoPilotActivityResponse,
    ReminderSettingsUpdate
)
from app.services.auto_pilot_scheduler import auto_pilot_scheduler


router = APIRouter()

# Auto-Pilot Plan Management
@router.post("/plans", response_model=AutoPilotPlanResponse)
async def create_auto_pilot_plan(
    plan_data: AutoPilotPlanCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create a new auto-pilot plan"""
    try:
        # Create the plan
        plan = AutoPilotPlan(
            name=plan_data.name,
            description=plan_data.description,
            frequency=plan_data.frequency,
            start_date=plan_data.start_date,
            end_date=plan_data.end_date,
            is_active=plan_data.is_active,
            question_rotation=plan_data.question_rotation,
            reminder_settings=plan_data.reminder_settings.dict(),
            distribution_channels=plan_data.distribution_channels,
            target_audience=plan_data.target_audience,
            max_responses=plan_data.max_responses,
            company_id=current_user.company_id,
            created_by=current_user.id
        )
        
        db.add(plan)
        db.commit()
        db.refresh(plan)
        
        return AutoPilotPlanResponse.from_orm(plan)
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create auto-pilot plan: {str(e)}"
        )

@router.get("/plans", response_model=List[AutoPilotPlanResponse])
async def get_auto_pilot_plans(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get all auto-pilot plans for the current user's company"""
    try:
        plans = db.query(AutoPilotPlan).filter(
            AutoPilotPlan.company_id == current_user.company_id
        ).all()
        
        return [AutoPilotPlanResponse.from_orm(plan) for plan in plans]
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get auto-pilot plans: {str(e)}"
        )

@router.get("/plans/{plan_id}", response_model=AutoPilotPlanResponse)
async def get_auto_pilot_plan(
    plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get a specific auto-pilot plan"""
    try:
        plan = db.query(AutoPilotPlan).filter(
            AutoPilotPlan.id == plan_id,
            AutoPilotPlan.company_id == current_user.company_id
        ).first()
        
        if not plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Auto-pilot plan not found"
            )
        
        return AutoPilotPlanResponse.from_orm(plan)
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get auto-pilot plan: {str(e)}"
        )

@router.put("/plans/{plan_id}", response_model=AutoPilotPlanResponse)
async def update_auto_pilot_plan(
    plan_id: int,
    plan_data: AutoPilotPlanUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update an auto-pilot plan"""
    try:
        plan = db.query(AutoPilotPlan).filter(
            AutoPilotPlan.id == plan_id,
            AutoPilotPlan.company_id == current_user.company_id
        ).first()
        
        if not plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Auto-pilot plan not found"
            )
        
        # Update plan fields
        for field, value in plan_data.dict(exclude_unset=True).items():
            if field == "reminder_settings":
                setattr(plan, field, value.dict())
            else:
                setattr(plan, field, value)
        
        plan.updated_at = datetime.now()
        db.commit()
        db.refresh(plan)
        
        return AutoPilotPlanResponse.from_orm(plan)
    except HTTPException:
        raise
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to update auto-pilot plan: {str(e)}"
        )

@router.delete("/plans/{plan_id}")
async def delete_auto_pilot_plan(
    plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Delete an auto-pilot plan"""
    try:
        plan = db.query(AutoPilotPlan).filter(
            AutoPilotPlan.id == plan_id,
            AutoPilotPlan.company_id == current_user.company_id
        ).first()
        
        if not plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Auto-pilot plan not found"
            )
        
        db.delete(plan)
        db.commit()
        
        return {"message": "Auto-pilot plan deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete auto-pilot plan: {str(e)}"
        )

@router.post("/plans/{plan_id}/activate", response_model=AutoPilotPlanResponse)
async def activate_auto_pilot_plan(
    plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Activate an auto-pilot plan"""
    try:
        plan = db.query(AutoPilotPlan).filter(
            AutoPilotPlan.id == plan_id,
            AutoPilotPlan.company_id == current_user.company_id
        ).first()
        
        if not plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Auto-pilot plan not found"
            )
        
        plan.is_active = True
        plan.updated_at = datetime.now()
        db.commit()
        db.refresh(plan)
        
        return AutoPilotPlanResponse.from_orm(plan)
    except HTTPException:
        raise
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to activate auto-pilot plan: {str(e)}"
        )

@router.post("/plans/{plan_id}/deactivate", response_model=AutoPilotPlanResponse)
async def deactivate_auto_pilot_plan(
    plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Deactivate an auto-pilot plan"""
    try:
        plan = db.query(AutoPilotPlan).filter(
            AutoPilotPlan.id == plan_id,
            AutoPilotPlan.company_id == current_user.company_id
        ).first()
        
        if not plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Auto-pilot plan not found"
            )
        
        plan.is_active = False
        plan.updated_at = datetime.now()
        db.commit()
        db.refresh(plan)
        
        return AutoPilotPlanResponse.from_orm(plan)
    except HTTPException:
        raise
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to deactivate auto-pilot plan: {str(e)}"
        )

# Auto-Pilot Survey Management
@router.get("/surveys", response_model=List[AutoPilotSurveyResponse])
async def get_scheduled_surveys(
    plan_id: Optional[int] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get scheduled surveys"""
    try:
        query = db.query(AutoPilotSurvey).join(AutoPilotPlan).filter(
            AutoPilotPlan.company_id == current_user.company_id
        )
        
        if plan_id:
            query = query.filter(AutoPilotSurvey.plan_id == plan_id)
        
        surveys = query.all()
        return [AutoPilotSurveyResponse.from_orm(survey) for survey in surveys]
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get scheduled surveys: {str(e)}"
        )

@router.get("/surveys/{survey_id}", response_model=AutoPilotSurveyResponse)
async def get_auto_pilot_survey(
    survey_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get a specific auto-pilot survey"""
    try:
        survey = db.query(AutoPilotSurvey).join(AutoPilotPlan).filter(
            AutoPilotSurvey.id == survey_id,
            AutoPilotPlan.company_id == current_user.company_id
        ).first()
        
        if not survey:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Auto-pilot survey not found"
            )
        
        return AutoPilotSurveyResponse.from_orm(survey)
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get auto-pilot survey: {str(e)}"
        )

@router.post("/surveys/{survey_id}/cancel")
async def cancel_auto_pilot_survey(
    survey_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Cancel an auto-pilot survey"""
    try:
        survey = db.query(AutoPilotSurvey).join(AutoPilotPlan).filter(
            AutoPilotSurvey.id == survey_id,
            AutoPilotPlan.company_id == current_user.company_id
        ).first()
        
        if not survey:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Auto-pilot survey not found"
            )
        
        survey.status = "cancelled"
        db.commit()
        
        return {"message": "Auto-pilot survey cancelled successfully"}
    except HTTPException:
        raise
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to cancel auto-pilot survey: {str(e)}"
        )

# Question Rotation
@router.get("/question-sets", response_model=List[dict])
async def get_question_sets(
    category: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get question sets for auto-pilot plans"""
    try:
        query = db.query(QuestionBank).filter(
            QuestionBank.metric.has(company_id=current_user.company_id)
        )
        
        if category:
            query = query.filter(QuestionBank.metric.has(category=category))
        
        question_sets = query.all()
        
        return [
            {
                "id": qs.id,
                "name": qs.name,
                "category": qs.category,
                "question_count": len(qs.questions),
                "questions": [
                    {
                        "id": q.id,
                        "text": q.text,
                        "category": q.category,
                        "order": q.order,
                        "required": q.required
                    }
                    for q in qs.questions
                ]
            }
            for qs in question_sets
        ]
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get question sets: {str(e)}"
        )

@router.post("/plans/{plan_id}/rotate-questions")
async def rotate_questions(
    plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Manually rotate questions for a plan"""
    try:
        plan = db.query(AutoPilotPlan).filter(
            AutoPilotPlan.id == plan_id,
            AutoPilotPlan.company_id == current_user.company_id
        ).first()
        
        if not plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Auto-pilot plan not found"
            )
        
        # Trigger question rotation
        await auto_pilot_scheduler.manual_trigger_survey(plan_id)
        
        return {"message": "Questions rotated successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to rotate questions: {str(e)}"
        )

# Reminder System
@router.post("/surveys/{survey_id}/send-reminder")
async def send_reminder(
    survey_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Manually send a reminder for a survey"""
    try:
        survey = db.query(AutoPilotSurvey).join(AutoPilotPlan).filter(
            AutoPilotSurvey.id == survey_id,
            AutoPilotPlan.company_id == current_user.company_id
        ).first()
        
        if not survey:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Auto-pilot survey not found"
            )
        
        # Trigger reminder
        await auto_pilot_scheduler.manual_trigger_reminder(survey_id)
        
        return {"message": "Reminder sent successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to send reminder: {str(e)}"
        )

@router.get("/surveys/{survey_id}/reminders")
async def get_reminder_history(
    survey_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get reminder history for a survey"""
    try:
        survey = db.query(AutoPilotSurvey).join(AutoPilotPlan).filter(
            AutoPilotSurvey.id == survey_id,
            AutoPilotPlan.company_id == current_user.company_id
        ).first()
        
        if not survey:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Auto-pilot survey not found"
            )
        
        # This would typically query a reminder history table
        # For now, return basic info
        return {
            "survey_id": survey_id,
            "reminder_count": survey.reminder_count,
            "last_reminder_date": survey.last_reminder_date,
            "max_reminders": survey.max_reminders
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get reminder history: {str(e)}"
        )

@router.put("/plans/{plan_id}/reminder-settings")
async def update_reminder_settings(
    plan_id: int,
    settings: ReminderSettingsUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update reminder settings for a plan"""
    try:
        plan = db.query(AutoPilotPlan).filter(
            AutoPilotPlan.id == plan_id,
            AutoPilotPlan.company_id == current_user.company_id
        ).first()
        
        if not plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Auto-pilot plan not found"
            )
        
        # Update reminder settings with new structure
        current_settings = plan.reminder_settings or {}
        settings_dict = settings.dict()
        
        # Ensure reminder days are properly set
        if 'reminderDays' not in settings_dict or not settings_dict['reminderDays']:
            settings_dict['reminderDays'] = [3, 7]  # Default reminder days
        
        # Ensure auto-close days are set
        if 'autoCloseAfterDays' not in settings_dict:
            settings_dict['autoCloseAfterDays'] = 10  # Default auto-close days
        
        current_settings.update(settings_dict)
        plan.reminder_settings = current_settings
        plan.updated_at = datetime.now()
        
        db.commit()
        db.refresh(plan)
        
        return AutoPilotPlanResponse.from_orm(plan)
    except HTTPException:
        raise
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to update reminder settings: {str(e)}"
        )

# Metrics and Analytics
@router.get("/metrics", response_model=AutoPilotMetricsResponse)
async def get_auto_pilot_metrics(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get auto-pilot metrics"""
    try:
        # Get plans for the company
        plans = db.query(AutoPilotPlan).filter(
            AutoPilotPlan.company_id == current_user.company_id
        ).all()
        
        # Get surveys for the company
        surveys = db.query(AutoPilotSurvey).join(AutoPilotPlan).filter(
            AutoPilotPlan.company_id == current_user.company_id
        ).all()
        
        # Calculate metrics
        total_surveys = len(surveys)
        active_plans = len([p for p in plans if p.is_active])
        
        # Calculate response rate (simplified)
        total_responses = sum(s.response_count for s in surveys)
        total_target = sum(s.target_count for s in surveys)
        average_response_rate = (total_responses / total_target * 100) if total_target > 0 else 0
        
        # Get next scheduled survey
        next_survey = None
        for plan in plans:
            if plan.is_active and plan.start_date <= datetime.now():
                next_date = plan.last_survey_date or plan.start_date
                if plan.frequency == "weekly":
                    next_date += timedelta(weeks=1)
                elif plan.frequency == "monthly":
                    next_date += timedelta(days=30)
                
                if not next_survey or next_date < next_survey:
                    next_survey = next_date
        
        return AutoPilotMetricsResponse(
            total_surveys=total_surveys,
            active_plans=active_plans,
            average_response_rate=round(average_response_rate, 1),
            total_responses=total_responses,
            next_scheduled_survey=next_survey,
            recent_activity=[]  # This would be populated from activity log
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get metrics: {str(e)}"
        )

@router.get("/activity", response_model=List[AutoPilotActivityResponse])
async def get_activity_log(
    plan_id: Optional[int] = None,
    limit: int = 10,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get activity log for auto-pilot plans"""
    try:
        # This would typically query an activity log table
        # For now, return mock data
        activities = []
        
        # Get recent surveys
        surveys = db.query(AutoPilotSurvey).join(AutoPilotPlan).filter(
            AutoPilotPlan.company_id == current_user.company_id
        ).order_by(AutoPilotSurvey.created_at.desc()).limit(limit).all()
        
        for survey in surveys:
            activities.append(AutoPilotActivityResponse(
                id=f"survey_{survey.id}",
                type="survey_sent",
                description=f"Survey '{survey.title}' sent",
                timestamp=survey.created_at,
                plan_id=survey.plan_id,
                survey_id=survey.id
            ))
        
        return activities[:limit]
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get activity log: {str(e)}"
        )
